File: src/jvmMain/java/jp/shanimnni/RCS380.java
package jp.shanimnni;

// from https://github.com/nearprosmith/java-rcs380-test

import org.apache.commons.lang3.ArrayUtils;

import javax.usb.*;
import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.util.Arrays;

public class RCS380 {
    static int VENDOR_ID = 0x054C;
    static int PRODUCT_ID = 0x06C3;
    private UsbPipe pipeIn;
    private UsbPipe pipeOut;
    private final UsbInterface iface;
    private final String manufacturer;
    private final String productName;

    public RCS380() throws UsbException, UnsupportedEncodingException, DeviceNotFoundException {
        UsbServices services = UsbHostManager.getUsbServices();

        UsbHub rootHub = services.getRootUsbHub();
        UsbDevice rcs380 = this.findDevice(rootHub, RCS380.VENDOR_ID, RCS380.PRODUCT_ID);

        if (rcs380 == null)
            throw new DeviceNotFoundException("Felica Device Not Found");

        this.manufacturer = rcs380.getManufacturerString();
        this.productName = rcs380.getProductString();

        UsbConfiguration configuration = (UsbConfiguration) rcs380.getUsbConfigurations().get(0);
        this.iface = (UsbInterface) configuration.getUsbInterfaces().get(0);


        UsbEndpoint endpointOut = null, endpointIn = null;
        for (int i = 0; i < iface.getUsbEndpoints().size(); i++) {
            byte endpointAddr = (byte) ((UsbEndpoint) (iface.getUsbEndpoints().get(i))).getUsbEndpointDescriptor().bEndpointAddress();
            if (((endpointAddr & 0x80) == 0x80)) {
                endpointIn = (UsbEndpoint) (iface.getUsbEndpoints().get(i));
            } else if ((endpointAddr & 0x80) == 0x00) {
                endpointOut = (UsbEndpoint) (iface.getUsbEndpoints().get(i))
                ;
            }
        }
        //0x02 : OUT, 0x081 IN
        endpointOut = iface.getUsbEndpoint((byte) 0x02);
        endpointIn = iface.getUsbEndpoint((byte) 0x81);

        this.pipeOut = endpointOut.getUsbPipe();
        this.pipeIn = endpointIn.getUsbPipe();
    }

    public void open() throws UsbException {
        this.iface.claim(usbInterface -> true);
        this.pipeIn.open();
        this.pipeOut.open();
        this.pipeOut.syncSubmit(Frame.ACK);
    }

    public void close() throws UsbException {
        this.pipeOut.abortAllSubmissions();
        sendCommand(Chipset.CMD_SWITCH_RF, new byte[]{0});
        this.pipeOut.syncSubmit(Frame.ACK);
        this.pipeIn.close();
        this.pipeOut.close();
        this.iface.release();
    }


    public ByteBuffer sendCommand(byte commandCode) {
        return this.sendCommand(this.pipeOut, this.pipeIn, commandCode, new byte[]{});
    }

    public ByteBuffer sendCommand(byte commandCode, byte[] commandData) {
        return this.sendCommand(this.pipeOut, this.pipeIn, commandCode, commandData);
    }

    public ByteBuffer sendCommand(UsbPipe pipeOut, UsbPipe pipeIn, byte commandCode) {
        return this.sendCommand(pipeOut, pipeIn, commandCode, new byte[]{});
    }

    public ByteBuffer sendCommand(UsbPipe pipeOut, UsbPipe pipeIn, byte commandCode, byte[] commandData) {
        Frame frame;
        byte[] data = new byte[255];
        try {
            frame = new Frame(ArrayUtils.addAll(new byte[]{(byte) 0xD6, commandCode}, commandData));
            pipeOut.syncSubmit(frame.frame);
            pipeIn.syncSubmit(data);
            frame = new Frame(data);

            if (frame.type.equals(Frame.TYPE_ACK)) {
                data = new byte[255];

                pipeIn.syncSubmit(data);
                frame = new Frame(data);
                if (frame.data[0] == (byte) 0xD7 && frame.data[1] == commandCode + 1) {
                    return ByteBuffer.wrap(Arrays.copyOfRange(frame.data, 2, frame.data.length));
                }
            } else {
                return null;
            }
        } catch (UsbException e) {
            e.printStackTrace();
            return null;
        }
        return null;
    }

    public UsbDevice findDevice(UsbHub hub, int vendorId, int productId) {
        for (Object obj : hub.getAttachedUsbDevices()) {
            if (!(obj instanceof UsbDevice))
                continue;
            UsbDevice device = (UsbDevice)obj;
            UsbDeviceDescriptor desc = device.getUsbDeviceDescriptor();

            if (desc.idVendor() == vendorId && desc.idProduct() == productId) return device;
            if (device.isUsbHub()) {
                device = findDevice((UsbHub) device, vendorId, productId);
                if (device != null) return device;
            }
        }
        return null;
    }

    public void setPipeIn(UsbPipe pipeIn) {
        this.pipeIn = pipeIn;
    }

    public void setPipeOut(UsbPipe pipeOut) {
        this.pipeOut = pipeOut;
    }

    public String getManufacturer() {
        return manufacturer;
    }

    public String getProductName() {
        return productName;
    }
}

File: src/jvmMain/kotlin/jp/shanimnni/RCS380.kt
package jp.shanimnni

import javax.usb.UsbPipe
import javax.usb.UsbInterface
import javax.usb.UsbException
import java.util.Arrays
import javax.usb.UsbHub
import javax.usb.UsbDevice
import javax.usb.UsbHostManager
import org.apache.commons.lang3.ArrayUtils
import java.nio.ByteBuffer
import javax.usb.UsbConfiguration
import javax.usb.UsbEndpoint
import kotlin.experimental.and

// from https://github.com/nearprosmith/java-rcs380-test
class RCS380 {
    private var pipeIn: UsbPipe
    private var pipeOut: UsbPipe
    private val iface: UsbInterface
    val manufacturer: String
    val productName: String
    @Throws(UsbException::class)
    fun open() {
        iface.claim { usbInterface: UsbInterface? -> true }
        pipeIn.open()
        pipeOut.open()
        pipeOut.syncSubmit(Frame.ACK)
    }

    @Throws(UsbException::class)
    fun close() {
        pipeOut.abortAllSubmissions()
        sendCommand(Chipset.CMD_SWITCH_RF, byteArrayOf(0))
        pipeOut.syncSubmit(Frame.ACK)
        pipeIn.close()
        pipeOut.close()
        iface.release()
    }

    fun sendCommand(commandCode: Byte): ByteBuffer? {
        return this.sendCommand(pipeOut, pipeIn, commandCode, byteArrayOf())
    }

    fun sendCommand(commandCode: Byte, commandData: ByteArray): ByteBuffer? {
        return this.sendCommand(pipeOut, pipeIn, commandCode, commandData)
    }

    @JvmOverloads
    fun sendCommand(
        pipeOut: UsbPipe,
        pipeIn: UsbPipe,
        commandCode: Byte,
        commandData: ByteArray = byteArrayOf()
    ): ByteBuffer? {
        var frame: Frame
        var data = ByteArray(255)
        try {
            frame = Frame(ArrayUtils.addAll(byteArrayOf(0xD6.toByte(), commandCode), *commandData))
            pipeOut.syncSubmit(frame.string)
            pipeIn.syncSubmit(data)
            frame = Frame(data)
            if(frame.type == Frame.TYPE_ACK) {
                data = ByteArray(255)
                pipeIn.syncSubmit(data)
                frame = Frame(data)
                if(frame.data[0] == 0xD7.toByte() && frame.data[1] == (commandCode + 1).toByte()) {
                    return ByteBuffer.wrap(Arrays.copyOfRange(frame.data, 2, frame.data.size))
                }
            } else {
                return null
            }
        } catch(e: UsbException) {
            e.printStackTrace()
            return null
        }
        return null
    }

    fun findDevice(hub: UsbHub?, vendorId: Int, productId: Int): UsbDevice? {
        for(obj in hub!!.attachedUsbDevices) {
            if(obj !is UsbDevice) continue
            var device = obj as UsbDevice?
            val desc = device!!.usbDeviceDescriptor
            if(desc.idVendor().toInt() == vendorId && desc.idProduct().toInt() == productId) return device
            if(device.isUsbHub) {
                device = findDevice(device as UsbHub?, vendorId, productId)
                if(device != null) return device
            }
        }
        return null
    }

    fun setPipeIn(pipeIn: UsbPipe) {
        this.pipeIn = pipeIn
    }

    fun setPipeOut(pipeOut: UsbPipe) {
        this.pipeOut = pipeOut
    }

    companion object {
        const val VENDOR_ID = 0x054C
        const val PRODUCT_ID = 0x06C3

        const val VENDOR_ID2 = 0x054C
        const val PRODUCT_ID2 = 0x06C1
    }

    init {
        val services = UsbHostManager.getUsbServices()
        val rootHub = services.rootUsbHub
        val rcs380 = findDevice(rootHub, VENDOR_ID, PRODUCT_ID)
            ?: findDevice(rootHub, VENDOR_ID2, PRODUCT_ID2)
            ?: throw DeviceNotFoundException("Felica Device Not Found")
        manufacturer = rcs380.manufacturerString
        productName = rcs380.productString
        val configuration = rcs380.usbConfigurations[0] as UsbConfiguration
        iface = configuration.usbInterfaces[0] as UsbInterface
        var endpointOut: UsbEndpoint? = null
        var endpointIn: UsbEndpoint? = null
        for(i in iface.usbEndpoints.indices) {
            val endpointAddr = (iface.usbEndpoints[i] as UsbEndpoint).usbEndpointDescriptor.bEndpointAddress()
            if((endpointAddr and (0x80).toByte()) == (0x80).toByte()) {
                endpointIn = iface.usbEndpoints[i] as UsbEndpoint?
            } else if((endpointAddr and (0x80).toByte()) == (0x00).toByte()) {
                endpointOut = iface.usbEndpoints[i] as UsbEndpoint?
            }
        }
        //0x02 : OUT, 0x081 IN
        endpointOut = iface.getUsbEndpoint(0x02.toByte())
        endpointIn = iface.getUsbEndpoint(0x81.toByte())
        pipeOut = endpointOut.usbPipe
        pipeIn = endpointIn.usbPipe
    }
}